source/daq/event_builder.cc

```diff
@@ -76,305 +77,457 @@ namespace psyllid
-                            t_write_flag->set_flag( false );
-                            LDEBUG( plog, "Event builder writing data to the output stream at index " << out_stream< 0 >().get_current_index() );
-                            if( ! out_stream< 0 >().set( midge::stream::s_run ) )
+                            LTRACE( plog, "No new trigger; Writing to from pretrig buffer only if buffer is full: " << f_pretrigger_buffer.full() );
+                            // contents of the buffer are the existing pretrigger plus the current trig id
+                            // only write out from the front of the buffer if the buffer is full; otherwise we're filling the buffer
+                            if( f_pretrigger_buffer.full() )
                             {
-                                LERROR( plog, "Exiting due to stream error" );
-                                break;
+                                LTRACE( plog, "Current state untriggered. Writing id "<<f_pretrigger_buffer.front()<<" as false");
+                                if( ! write_output_from_ptbuff_front( false, t_write_flag ) )
+                                {
+                                    break;
+                                }
+                                // pretrigger buffer is full - 1
                             }
-
-                            f_pretrigger_buffer.pop_front();
-
-                            continue;
                         }
                     }
-                    if( f_state == state_t::untriggered_nopt )
+                    else if (f_state == state_t::collecting_triggers)
                     {
-                        if( t_current_trig_flag )
+                        LTRACE( plog, "Currently in waiting state" );
+                        if (t_current_trig_flag)
                         {
-                            LDEBUG( plog, "Was: untriggered - no pt; Now: new trigger" );
-
-                            t_write_flag->set_id( t_trigger_flag->get_id() );
-                            t_write_flag->set_flag( true );
-                            LDEBUG( plog, "Event builder writing data to the output stream at index " << out_stream< 0 >().get_current_index() );
-                            if( ! out_stream< 0 >().set( midge::stream::s_run ) )
+                            t_trigger_count++;
+                            LINFO (plog, "Got another trigger: "<<t_trigger_count<<" need N triggers: "<<f_n_triggers);
+                            if (t_trigger_count == f_n_triggers)
                             {
-                                LERROR( plog, "Exiting due to stream error" );
-                                break;
+                                t_trigger_count = 0;
+                                // flush the pretrigger buffer as true, which includes the current trig id
+                                while( ! f_pretrigger_buffer.empty() )
```

Comment:
I don't entirely understand this part.  Why do we write out both the pretrigger buffer and the skip buffer?

Reply:
When the event builder switches from untriggered to collecting it doesnt write out the pretrigger buffer yet. In the collecting state the skip buffer is filled. When enough triggers arrive both buffers are emptied as true otherwise as false.

Reply:
Ok, I see what you're saying.  Doesn't that lead to duplication where the two buffers overlap?

Reply:
they dont overlap. the pretrigger buffer doesnt get new ids in collecting state. 
and when not enough triggers arrive and the pretrigger buffer is emptied as false and refilled with ids from the skip buffer. if the skip buffer size is larger it is emptied as false untill the remaining ids all fit in the pretrigger buffer.

Reply:
Ah, ok.  That makes sense.  Thanks for the help understanding that.